subprojects/language-java/src/main/java/org/gradle/api/tasks/compile/JavaCompile.java

```diff
@@ -202,7 +202,8 @@ protected Deleter getDeleter() {
 
     @Nested
     protected JavaPlatform getPlatform() {
-        return new DefaultJavaPlatform(JavaVersion.toVersion(getTargetCompatibility()));
+        JavaVersion javaVersion = JavaVersion.toVersion(getTargetCompatibility());
```

Comment:
This is one of the reasons that I backed out of visiting the nested output properties initially, as calling the getter exposes all kinds of bad behaviours in plugins that previously weren't exposed previously. Even if we catch the exceptions and report on them, we are now paying the cost of throwing a bunch of exceptions at configuration time, which can be really expensive. And just propagating the exceptions would be a nasty breaking change.

I would look at instead attaching the producer task lazily, when the getter is called, instead of doing it eagerly when the task is created.

Another option might be to use annotations and/or types that better reflect the role that the property plays, so that we know whether or not the property is relevant for attaching outputs. For example, the things that are fixed below actually represent tools that the task use, rather than an arbitrary bucket of beans that might define some outputs. If we better modelled tools (say as build services or as an explicit first class concept), we would not need to query these properties until we need to calculate the task dependencies.

Regardless of how we decide to proceed with this, we should add some test coverage where a custom task type has a getter that fails when called, until some other property has been configured.